crowbar_framework/app/models/service_object.rb

```diff
@@ -1205,31 +1306,40 @@ def apply_role(role, inst, in_queue)
           end
 
           old_nodes.each do |node_name|
-            # Don't add deleted nodes to the run order
+            # Don't add deleted nodes to the run order, they clearly won't have the old role
             if NodeObject.find_node_by_name(node_name).nil?
               @logger.debug "skipping deleted node #{node_name}"
               next
             end
 
+            # An old node that is not in the new deployment, drop it
             if new_nodes.nil? or !new_nodes.include?(node_name)
               @logger.debug "remove node #{node_name}"
               pending_node_actions[node_name] = { :remove => [], :add => [] } if pending_node_actions[node_name].nil?
+
               pending_node_actions[node_name][:remove] << role_name
+
+              # FIXME: are remove_roles a way how to tell us to do some post-cleanup?
               pending_node_actions[node_name][:add] << elem_remove unless elem_remove.nil?
+
               nodes_in_batch << node_name
             end
           end
         end
 
+        # FIXME: what does it mean if new_nodes is nil?
         unless new_nodes.nil?
           new_nodes.each do |node_name|
             # Don't add deleted nodes to the run order
+            # FIXME: if we want to deploy on a deleted node, shouldn't we just bail out instead?
```

Comment:
IIRC, this got added for the barclamps where all nodes are used (for instance, provisioner, logging, dns, ntp); so that we don't fail too easily when a node got forgotten.

It's kind of a ugly workaround for the fact that we don't properly handle forgotten node + for the fact that we don't have some alias that be used to assign all existing nodes to a role (which would be an improvement over the requirement to list all nodes).